File: gateway/builder.go
package gateway

import (
	"fmt"
	"strings"

	"github.com/ZacharyZcR/Kelper/core/llm"
	"github.com/getkin/kin-openapi/openapi3"
)

type AuthConfig struct {
	Type       string
	HeaderName string
}

type Endpoint struct {
	TargetName string
	BaseURL    string
	Method     string
	Path       string
	Auth       AuthConfig
	Params     []ParamInfo
	HasBody    bool
}

type ParamInfo struct {
	Name     string
	In       string // path, query, header
	Required bool
}

func BuildTools(doc *openapi3.T, targetName, baseURL string, auth AuthConfig) ([]llm.Tool, map[string]*Endpoint) {
	var tools []llm.Tool
	endpoints := make(map[string]*Endpoint)

	if doc.Paths == nil {
		return tools, endpoints
	}

	for path, pathItem := range doc.Paths.Map() {
		for method, op := range pathItem.Operations() {
			if op == nil {
				continue
			}

			opID := op.OperationID
			if opID == "" {
				opID = generateOpID(method, path)
			}
			toolName := targetName + "__" + opID

			description := op.Summary
			if description == "" {
				description = op.Description
			}
			if description == "" {
				description = fmt.Sprintf("%s %s", strings.ToUpper(method), path)
			}

			params, paramInfos := buildParams(op)

			tool := llm.Tool{
				Type: "function",
				Function: llm.ToolFunction{
					Name:        toolName,
					Description: description,
					Parameters:  params,
				},
			}

			endpoint := &Endpoint{
				TargetName: targetName,
				BaseURL:    baseURL,
				Method:     strings.ToUpper(method),
				Path:       path,
				Auth:       auth,
				Params:     paramInfos,
				HasBody:    op.RequestBody != nil,
			}

			tools = append(tools, tool)
			endpoints[toolName] = endpoint
		}
	}

	return tools, endpoints
}

func buildParams(op *openapi3.Operation) (map[string]interface{}, []ParamInfo) {
	properties := map[string]interface{}{}
	var required []string
	var paramInfos []ParamInfo

	for _, paramRef := range op.Parameters {
		if paramRef == nil || paramRef.Value == nil {
			continue
		}
		p := paramRef.Value
		prop := map[string]interface{}{"type": "string"}
		if p.Schema != nil && p.Schema.Value != nil {
			prop = schemaToMap(p.Schema.Value)
		}
		if p.Description != "" {
			prop["description"] = fmt.Sprintf("%s (%s)", p.Description, p.In)
		}
		properties[p.Name] = prop
		if p.Required {
			required = append(required, p.Name)
		}
		paramInfos = append(paramInfos, ParamInfo{Name: p.Name, In: p.In, Required: p.Required})
	}

	if op.RequestBody != nil && op.RequestBody.Value != nil {
		for _, mediaType := range op.RequestBody.Value.Content {
			if mediaType.Schema != nil && mediaType.Schema.Value != nil {
				properties["body"] = schemaToMap(mediaType.Schema.Value)
				if op.RequestBody.Value.Required {
					required = append(required, "body")
				}
			}
			break
		}
	}

	schema := map[string]interface{}{
		"type":       "object",
		"properties": properties,
	}
	if len(required) > 0 {
		schema["required"] = required
	}

	return schema, paramInfos
}

func schemaToMap(s *openapi3.Schema) map[string]interface{} {
	if s == nil {
		return map[string]interface{}{"type": "string"}
	}

	m := map[string]interface{}{}

	if s.Type != nil {
		types := s.Type.Slice()
		if len(types) == 1 {
			m["type"] = types[0]
		} else if len(types) > 1 {
			m["type"] = types
		}
	}
	if s.Description != "" {
		m["description"] = s.Description
	}
	if len(s.Enum) > 0 {
		m["enum"] = s.Enum
	}
	if len(s.Properties) > 0 {
		props := map[string]interface{}{}
		for name, ref := range s.Properties {
			if ref != nil && ref.Value != nil {
				props[name] = schemaToMap(ref.Value)
			}
		}
		m["properties"] = props
	}
	if len(s.Required) > 0 {
		m["required"] = s.Required
	}
	if s.Items != nil && s.Items.Value != nil {
		m["items"] = schemaToMap(s.Items.Value)
	}

	return m
}

func generateOpID(method, path string) string {
	path = strings.ReplaceAll(path, "{", "")
	path = strings.ReplaceAll(path, "}", "")
	path = strings.ReplaceAll(path, "/", "_")
	path = strings.Trim(path, "_")
	return strings.ToLower(method) + "_" + path
}

File: sdk/engine/src/gateway/builder.ts
import type { Tool, Endpoint, ParamInfo, AuthConfig } from '../types.js';

const RE_INVALID = /[^a-zA-Z0-9_.\-:]/g;

function sanitizeToolName(name: string): string {
  name = name.replace(RE_INVALID, '_');
  if (!name) return '_';
  if (!/^[a-zA-Z_]/.test(name)) name = '_' + name;
  if (name.length > 64) name = name.slice(0, 64);
  return name;
}

function generateOpID(method: string, path: string): string {
  const cleaned = path.replace(/[{}]/g, '').replace(/\//g, '_').replace(/^_|_$/g, '');
  return method.toLowerCase() + '_' + cleaned;
}

/**
 * Resolve a $ref pointer like "#/components/schemas/Pet" against the root doc.
 */
function resolveRef(doc: Record<string, unknown>, ref: string): Record<string, unknown> | null {
  if (!ref.startsWith('#/')) return null;
  const parts = ref.slice(2).split('/');
  let cur: unknown = doc;
  for (const p of parts) {
    if (cur == null || typeof cur !== 'object') return null;
    cur = (cur as Record<string, unknown>)[p];
  }
  return (typeof cur === 'object' && cur !== null) ? cur as Record<string, unknown> : null;
}

/**
 * Convert an OpenAPI schema node into a plain JSON Schema object for LLM tool parameters.
 * Handles $ref resolution.
 */
function schemaToMap(doc: Record<string, unknown>, schema: Record<string, unknown>): Record<string, unknown> {
  if (!schema) return { type: 'string' };

  // Resolve $ref
  if (typeof schema['$ref'] === 'string') {
    const resolved = resolveRef(doc, schema['$ref']);
    if (resolved) return schemaToMap(doc, resolved);
    return { type: 'string' };
  }

  const m: Record<string, unknown> = {};

  if (schema.type) m.type = schema.type;
  if (schema.description) m.description = schema.description;
  if (Array.isArray(schema.enum) && schema.enum.length) m.enum = schema.enum;

  if (schema.properties && typeof schema.properties === 'object') {
    const props: Record<string, unknown> = {};
    for (const [name, ref] of Object.entries(schema.properties as Record<string, unknown>)) {
      if (ref && typeof ref === 'object') props[name] = schemaToMap(doc, ref as Record<string, unknown>);
    }
    m.properties = props;
  }

  if (Array.isArray(schema.required) && schema.required.length) m.required = schema.required;
  if (schema.items && typeof schema.items === 'object') {
    m.items = schemaToMap(doc, schema.items as Record<string, unknown>);
  }

  return m;
}

interface BuildResult {
  tools: Tool[];
  endpoints: Map<string, Endpoint>;
}

/**
 * Build LLM Tool definitions + Endpoint map from an OpenAPI spec object.
 * Mirrors Go's gateway.BuildTools.
 */
export function buildTools(
  doc: Record<string, unknown>,
  targetName: string,
  baseURL: string,
  auth: AuthConfig,
): BuildResult {
  const tools: Tool[] = [];
  const endpoints = new Map<string, Endpoint>();

  const paths = doc.paths as Record<string, Record<string, unknown>> | undefined;
  if (!paths) return { tools, endpoints };

  for (const [path, pathItem] of Object.entries(paths)) {
    if (!pathItem || typeof pathItem !== 'object') continue;

    const methods = ['get', 'post', 'put', 'delete', 'patch', 'head', 'options'] as const;
    for (const method of methods) {
      const op = pathItem[method] as Record<string, unknown> | undefined;
      if (!op) continue;

      const opID = (op.operationId as string) || generateOpID(method, path);

      let sanitizedTarget = sanitizeToolName(targetName);
      if (sanitizedTarget.replace(/_/g, '') === '') sanitizedTarget = 'target';

      const toolName = sanitizeToolName(sanitizedTarget + '__' + opID);

      const description =
        (op.summary as string) ||
        (op.description as string) ||
        `${method.toUpperCase()} ${path}`;

      const { schema, paramInfos } = buildParams(doc, op);

      tools.push({
        type: 'function',
        function: { name: toolName, description, parameters: schema },
      });

      endpoints.set(toolName, {
        targetName: sanitizedTarget,
        targetDisplayName: targetName,
        baseURL,
        method: method.toUpperCase(),
        path,
        auth,
        params: paramInfos,
        hasBody: op.requestBody != null,
      });
    }
  }

  return { tools, endpoints };
}

function buildParams(
  doc: Record<string, unknown>,
  op: Record<string, unknown>,
): { schema: Record<string, unknown>; paramInfos: ParamInfo[] } {
  const properties: Record<string, unknown> = {};
  const required: string[] = [];
  const paramInfos: ParamInfo[] = [];

  const params = op.parameters as Array<Record<string, unknown>> | undefined;
  if (params) {
    for (const paramOrRef of params) {
      // Resolve parameter $ref
      const p = (typeof paramOrRef['$ref'] === 'string')
        ? resolveRef(doc, paramOrRef['$ref'] as string) ?? paramOrRef
        : paramOrRef;
      if (!p || !p.name) continue;

      let prop: Record<string, unknown> = { type: 'string' };
      const schema = p.schema as Record<string, unknown> | undefined;
      if (schema) prop = schemaToMap(doc, schema);
      if (p.description) prop.description = `${p.description} (${p.in})`;

      properties[p.name as string] = prop;
      if (p.required) required.push(p.name as string);
      paramInfos.push({
        name: p.name as string,
        in: p.in as 'path' | 'query' | 'header',
        required: !!p.required,
      });
    }
  }

  // Request body
  const requestBody = op.requestBody as Record<string, unknown> | undefined;
  if (requestBody) {
    // Resolve requestBody $ref
    const rb = (typeof requestBody['$ref'] === 'string')
      ? resolveRef(doc, requestBody['$ref'] as string) ?? requestBody
      : requestBody;
    const content = rb.content as Record<string, Record<string, unknown>> | undefined;
    if (content) {
      for (const mediaType of Object.values(content)) {
        const schema = mediaType.schema as Record<string, unknown> | undefined;
        if (schema) {
          properties.body = schemaToMap(doc, schema);
          if (rb.required) required.push('body');
        }
        break; // take first
      }
    }
  }

  const schema: Record<string, unknown> = { type: 'object', properties };
  if (required.length) schema.required = required;

  return { schema, paramInfos };
}
